presto-main/src/main/java/com/facebook/presto/memory/ClusterMemoryManager.java

```diff
@@ -107,7 +121,63 @@ public void process(Iterable<QueryExecution> queries)
         return ImmutableMap.copyOf(pools);
     }
 
-    private void updateNodes()
+    private MemoryPoolAssignmentsRequest updateAssignments(Iterable<QueryExecution> queries)
+    {
+        ClusterMemoryPool reservedPool = pools.get(RESERVED_POOL);
+        ClusterMemoryPool generalPool = pools.get(GENERAL_POOL);
+        long version = memoryPoolAssignmentsVersion.incrementAndGet();
+        // Check that all previous assignments have propagated to the visible nodes. This doesn't account for temporary network issues,
+        // and is more of a safety check than a guarantee
+        if (reservedPool != null && generalPool != null && allAssignmentsHavePropagated(queries)) {
+            if (reservedPool.getQueries() == 0 && generalPool.getBlockedNodes() > 0) {
+                QueryExecution biggestQuery = null;
+                long maxMemory = -1;
+                for (QueryExecution queryExecution : queries) {
+                    long bytesUsed = queryExecution.getTotalMemoryReservation();
+                    if (bytesUsed > maxMemory) {
+                        biggestQuery = queryExecution;
+                        maxMemory = bytesUsed;
+                    }
+                }
+                for (QueryExecution queryExecution : queries) {
+                    if (queryExecution.getQueryId().equals(biggestQuery.getQueryId())) {
+                        queryExecution.setMemoryPool(new VersionedMemoryPoolId(RESERVED_POOL, version));
+                    }
+                    else {
+                        queryExecution.setMemoryPool(new VersionedMemoryPoolId(GENERAL_POOL, version));
```

Comment:
Is this really needed?  Don't you just want to find the query that needs to be move and move only that one?